trl/trainer/grpo_trainer.py

```diff
@@ -1244,13 +1265,43 @@ def _generate_and_score_completions(
         for i, name in enumerate(self.reward_func_names):
             self._textual_logs["rewards"][name].extend(rewards_per_func[:, i].tolist())
 
+        # Unpad the prompt and completion ids to optimise memory: https://github.com/huggingface/trl/pull/3495
+        unpadded_prompt_ids = []
+        unpadded_completion_ids = []
+        unpadded_per_token_logps = []
+        prompt_ids_length = prompt_ids.size(1)
+        logp_iter = attention_mask if old_per_token_logps is None else old_per_token_logps  # dummy iterator
+        # get the start and end indices of the attention_mask
+        for p_ids, c_ids, old_logps, mask in zip(prompt_ids, completion_ids, logp_iter, attention_mask):
+            indices = torch.where(mask == 1)[0]
+
+            if len(indices) > 0:
+                start = indices[0]
+                end = indices[-1] + 1
+                if prompt_ids_length >= end:
+                    raise ValueError(
+                        f"End index {end} exceeds prompt_ids_length {prompt_ids_length}. "
+                        "This can happen if the attention mask is not correctly set."
+                    )
+                # prompt ids were left padded
+                unpadded_prompt_ids.append(p_ids[start:prompt_ids_length])
+                # completion ids were right padded
+                unpadded_completion_ids.append(c_ids[: end - prompt_ids_length])
+                if mask[start:end].sum() != end - start:
+                    raise ValueError(f"Attention mask from {start} to {end} does not match the expected length. ")
+                if old_per_token_logps is not None:
+                    unpadded_per_token_logps.append(old_logps[: end - prompt_ids_length])
+            else:
+                # case where the attention mask is all zeros, e.g. when mask_truncated_completions is enabled
```

Comment:
Can you double-check this? In case of masked truncated completion, I think only the completion mask is truncated. You'd have something like:

```python
prompt_ids = [0, 0, 1, 2]
prompt_mask = [0, 0, 1, 1]
completion_ids = [3, 4, 5, 0]
completion_mask = [0, 0, 0, 0]
```

hence:

```python
attention_mask = [0, 0, 1, 1, 0, 0, 0, 0]
```